babel/number_spelling.py

```diff
@@ -0,0 +1,549 @@
+        print('__uni__', s)
+
+        return self.prefix + s + self.suffix
+
+    def __str__(self):
+        """
+        This is necessary to support both Python 2 and 3
+        """
+        if sys.version < '3':
+            return unicode(self).encode('utf-8')
+        else:
+            return self.__unicode__()
+
+
+class NumberContext(ContextBase):
+    """
+    Represents a number and its internal structure.
+
+    The number is separated into integer and fraction parts
+    and the fraction part is also rounded if necessary.
+    """
+    def __init__(self, value, integer, fraction, rounded, negative):
+        ContextBase.__init__(self)
+        self.value = value
+        self.integer = SideContext(integer, self)
+        self.fraction = SideContext(fraction, self)
+        self.rounded = rounded
+        self.negative = negative
+
+    def __iter__(self):
+        yield self.fraction
+        yield self.integer
+
+    def __len__(self):
+        return len(self.integer) + len(self.fraction)
+
+    @property
+    def last_nonzero(self):
+        """
+        The first non-zero digit from the right
+        """
+        for side in self:
+            for group in side:
+                for digit in group:
+                    if digit.value != 0:
+                        return digit
+
+
+class SideContext(ContextBase):
+    """
+    Represent one side of the decimal point
+    """
+    def __init__(self, groups, number):
+        ContextBase.__init__(self)
+        self.number = number
+        self.groups = tuple(GroupContext(digits, number, self, i) for i, digits in enumerate(groups))
+        # calculate the value of the side the hard way (might use number.value instead)
+        e, s = 0, 0
+        for g in self.groups:
+            s += g.value*10**e
+            e += len(g)
+        self.value = s
+
+    def __iter__(self):
+        for group in self.groups:
+            yield group
+
+    def __len__(self):
+        return sum(len(g) for g in self)
+
+
+class GroupContext(ContextBase):
+    """
+    Represent a group of digits inside a number
+
+    Digits inside the group are indexed from the right.
+    """
+    def __init__(self, digits, number, side, index):
+        ContextBase.__init__(self)
+        self.digits = tuple(DigitContext(v, number, side, self, i) for i,v in enumerate(digits))
+        self.number = number
+        self.side = side
+        self.index = index
+        self.value = sum(d.value*dec(10)**e for e,d in enumerate(self.digits)) # might use int
+
+    def __len__(self):
+        return len(self.digits)
+
+    def __iter__(self):
+        for digit in self.digits:
+            yield digit
+
+
+class DigitContext(ContextBase):
+    """
+    Represent one digit of a number and its context
+    """
+    def __init__(self, value, number, side, group, index):
+        ContextBase.__init__(self)
```

Comment:
Ok, I will incorporate all the changes. Should I make a new pull request or
still modify the current commit.

I have nothing against super, I decided to use the simple way to make it
simple, but I could add super no problem.

I also add the factoring of the locale specific code and some loading
mechanism.

I'll leave out decimals global context.

Get rid of unused files.

Get back to you as soon as I have the time to finish all that.

Szabolcs
On 4 Aug 2015 22:09, "Isaac Jurado" <notifications@github.com> wrote:

> In babel/number_spelling.py
> <https://github.com/mitsuhiko/babel/pull/114#discussion_r36235049>:
>
> > +        self.value = sum(d.value*dec(10)**e for e,d in enumerate(self.digits)) # might use int
> > +
> > +    def __len__(self):
> > +        return len(self.digits)
> > +
> > +    def __iter__(self):
> > +        for digit in self.digits:
> > +            yield digit
> > +
> > +
> > +class DigitContext(ContextBase):
> > +    """
> > +    Represent one digit of a number and its context
> > +    """
> > +    def __init__(self, value, number, side, group, index):
> > +        ContextBase.__init__(self)
>
> Just for the record. super() without arguments only works in Python 3.
> Python 2 requires two arguments for super().
>
> Highly recommended talk: Super considered super! by Raymond Hettinger
> <https://www.youtube.com/watch?v=EiOglTERPEo>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mitsuhiko/babel/pull/114/files#r36235049>.
>